### ISSUE TITLE
Running DB Install as Part of CI // RSA publickey Error When Building New Images 

### ISSUE THREAD

**Issue by zfalen** [directus/docker#70]
Hello! This is a two-part issue, though solving either will likely solve the other.

I am attempting to get the dist versions of Directus Docker images working with a Jenkins/Rancher CI pipeline. As currently configured, it's a pretty big pain to shell into the host each time to run the `docker-compose` install script every time we spin up a new app that will use Directus as a CMS layer. 

**So, Issue #1:** we'd like to find a way to run the install script when the container is stood up initially, _after_ the SQL container is healthy. See the following links for a few ways to do this: 
https://docs.docker.com/compose/startup-order/
https://stackoverflow.com/questions/42117362/how-to-execute-scripts-after-docker-compose-up

Depending on the version of Rancher, there are also some ways to explicitly tell a service in `docker-compose.yml` not to start until it's `links` are in a "healthy" state, but the links above show a more up-to-date method via entrypoint. 

Currently, I don't see any way to make this happen _(although, looking at the repo in `images/base/api/..../api-entrypoint` I do see an option for a `DIRECTUS_INSTALL="auto"` env variable that would run the install script)_. However, setting that flag _also_ triggers an `exit $?` on line 74, so the container would need to be restarted, which seems to trigger the install script again, and fails because it's already been run once before. _(So, something like `restart: always` in `docker-compose.yml` also does not work 😁)_

So, in _theory_ I can break this part into two sections, one which looks like this: 
```
# Allow ENV to auto install the DB and continue living 
if [ "${DIRECTUS_INSTALL}" == "auto" ]; then
  install
fi
```
This is an untested assumption, and likely requires an addition to the entrypoint script that does something similar to `touch .initialized` described in the second link above to flag that the DB is, in fact, installed if the container restarts itself, but I _think_ it would work?

_**But**_ this leads me to....

**Issue #2:** I cannot get the build script to work due to some SSH / RSA key error. 

This appears to be happening inside one of the containers that gets stood up as part of a RUN command doing `composer` tasks. Raw error:
```
Loading composer repositories with package information


  [RuntimeException]
  Failed to clone https://github.com/directus/zend-db via https, ssh protocols, aborting.

  - https://github.com/directus/zend-db
    Cloning into bare repository '/root/.composer/cache/vcs/https---github.com-directus-zend-db'...
    fatal: unable to access 'https://github.com/directus/zend-db/': server certificate verification failed. CAfile: none CRLfile: none

  - git@github.com:directus/zend-db
    Cloning into bare repository '/root/.composer/cache/vcs/https---github.com-directus-zend-db'...
    Warning: Permanently added the RSA host key for IP address '192.30.253.113' to the list of known hosts.
    git@github.com: Permission denied (publickey).
    fatal: Could not read from remote repository.

    Please make sure you have the correct access rights
    and the repository exists.
```

The build command I am running is: `./bin/build --type dist --project api --kind apache`

I've verified that I am running `Bash v5.0.11(1)-release`, and my personal private/public RSA keys are configured with GitHub on my machine _(plus I can clone the repo in question to my machine)._ So I think it is an issue inside the docker container as it gets built?

**Comment by zfalen** [comment]
To be clear, if I can get unblocked on the build issue I can play with the API install scripts and find /contribute a workable solution for other people who are interested in a solution for the first issue 😁Looks like most of it is already done anyway. I like the ENV variable approach that can live in `docker-compose.yml`. 

But maybe it already exists and I am just missing something.

**Comment by WoLfulus** [comment]
I've decided not to implement auto installation feature and use user-triggered installation instead because it's too risky to assume and detect scenarios where the installation needs to occur. It is quite common to have multiple concurrent containers (which they all might want install themselves) and having to mount a volume only for a `.initialized` file in your case. We'd like to keep containers as stateless as possible (even though we have support for storing stuff in volumes, we suggest using s3 instead in this case).

I've also thought about displaying a setup dialog, but we ended up implementing the CLI command solution instead, which is easy to automate in k8s for example (see https://github.com/directus/docker/issues/3#issuecomment-528553750), 

There should also be a command for db migration in the future, but it's not implemented yet.

If you really want to do auto installation, things I can suggest are: 

- Having a fork of the api (and creating a Dockerfile in the root - build script will take it as priority)
  `/Dockerfile`
  ```
  ARG BASE_IMAGE
  FROM ${BASE_IMAGE}
  COPY custom.sh /custom
  RUN chmod +x /custom
  ENTRYPOINT ["/custom"]
  ```
  `/custom.sh`
  ```
  #!/usr/bin/env bash

  # custom auto install logic here  

  exec api-entrypoint $@
  ```

- Cloning the api and adding those two files files as an additional step before building


The advantage of using the fork, is that the build script allows you to pass `--repository <user>/<repo>` for github repos, or `--repository /some/local/path` as the local source repository, and combining with `--ref <master|commit_sha|tag>` you can build from an specific commit.


As for the second issue, I'm unable to reproduce it here.

Let me know what you think.

**Comment by zfalen** [comment]
Interesting. I agree that it's tough to tell when the install needs to occur, especially without a `volume` solution of some kind - although we'll probably go that route in the short / medium term to persist DB data. I like the S3 route. 

Still, seems like something that is _relatively_ easy to expose to users with minimal changes to the way things are currently wired, assuming the user is aware of the risks / need for making the container more "stateful" with a `.initialized` file or something. But not a default option. Our heads are in the same spot on how it would implement, I think, anyways. 

I'll go ahead and try forking it as you suggested. Accomplishes what I'm trying to do and the extra CLI args make it super easy to point at. Nice! 

However, I'm still getting the RSA errors... I am not sure why. I'll keep playing with it. 

_(Maybe helpful)_ **Context:**
We're still getting familiar with Directus as a CMS layer, but basically we're an agency that's building / managing both a suite of internal tools as well as client apps / microsites / etc where CMS is an important ask from the users. In most client cases, the Suite makes sense - but being able to decouple / link lots of API instances to one App is really useful for the internal tools, and sometimes for bigger client ecosystems. 

We Dockerize everything anyways, and our CI is set up in such a way it would be really convenient to just add the Directus API images and know they'll spin up and initialize along with the rest of the app on the first push. Clean, simple, new tool added to the ecosystem and we can plug into it on a single App. Need a CMS-controllable data layer? Plug it in. Easy. 

Of course we can manually shell in and run the install command every time on the host after it gets up, but it's an extra step outside of how our devs are used to using the CI workflow. 

Where we're struggling is running the install as a `docker-compose` command since Rancher 1.6 doesn't seem to let us easily automate running a DC script _after_ containers are up and the official Docker position on that functionality appears to be "it's a dumb idea even though everyone wants it" _(https://github.com/docker/compose/issues/1809)_

The general _(docker team)_ consensus on that thread seems to be modifying existing Dockerfiles with an entrypoint to run a script within a specific container when it stands up _after_ some other container passes a health check, or something, is the right way to approach things. But it just seems silly when you give us a nice `docker-compose` command to run 😄This may be way easier on the 2+ version of Rancher using k8s but I am not sure. We're not set up for it yet.

**Comment by zfalen** [comment]
**_TLDR:_ Disabling a dumb corporate firewall fixed the RSA errors.**

**Comment by zfalen** [comment]
@WoLfulus heads up that some of the `-d` checks for existing directories in the build script should have quotes around them to deal with weird people like me who have spaces in some directory names 😁

E.G. line 212
`if ! [ -d "$root/images/base/${project}/" ]; then`
instead of 
`if ! [ -d $root/images/base/${project}/ ]; then`

**Comment by zfalen** [comment]
@WoLfulus I have it working locally and will edit this comment as soon as we get the fork, etc set up correctly.

**Comment by zfalen** [comment]
@WoLfulus side question - is there a `docker-compose` example that shows how to set up a single App instance to manage multiple API instances? 

In the example I pulled from this repo _(`simple`, I think)_ the App is set up as follows:
```
  # Directus APP instance.
  # This is our graphical management interface.
  directus-app:
    image: directus/app:v7-node
    ports:
      - "8090:80"
    environment:
      # Endpoints
      DIRECTUS_APP_TITLE: "Heat Team Cards"
      DIRECTUS_APP_URL: "http://localhost:8080/_/"
```

Can I pass, say, an array to these options and have it set up the relevant config or will I need to shell in and update the config as described here:
https://docs.directus.io/advanced/app/configuration.html#allowed-apis

**Comment by WoLfulus** [comment]
Sorry the delay! But yeah, that's how you do it!

**Comment by zfalen** [comment]
No worries. I am continuing to tweak things as we go along. Circle back whenever you have the time and I can run you over our solution 😄

**Comment by zfalen** [comment]
@WoLfulus - little update here. I manage to get things stable by tweaking the `api-entrypoint` to do this:

```
#
# Bootstrap
#

     echo "Waiting for MySQL DB to stand up..."
     while ! mysqladmin ping -h mysql --user=${DIRECTUS_DATABASE_USERNAME} -- 
     password=${DIRECTUS_DATABASE_PASSWORD} --silent ; do 
        sleep 10 
     done
     echo "MySQL DB is up - moving on"

     # .... Some stuff happens

     # Allow ENV to auto install the DB and continue living 
     if [ "${DIRECTUS_INSTALL}" == "auto" ]; then
          if ! [ -f "$PWD/installConfig/.dbInstalled" ]; then
              install
          fi
     fi

     # .... Some stuff happens, install() is called

     /var/www/bin/directus install:database
  
    # CHECK FOR SPECIAL ERROR CODE THAT ONLY THROWS IF DB IS ALREADY INSTALLED
    if [ "$?" == "3" ]; then
      if [ "${DIRECTUS_INSTALL}" == "auto" ] && ! [ -f "$PWD/installConfig/.dbInstalled" ]; then
        # ERROR INDICATES DB IS ALREADY INITIALIZED
        echo ""
        echo "DB EXISTS BUT NOT THE FLAG FILE... RE-CREATING!"
        echo ""
        
        # CONTAINER PROBABLY RESTARTED AND THE VOLUME LOST THE FLAG FILE
        # JUST RECREATE THE FILE AND MOVE ON
        touch ./installConfig/.dbInstalled
        return 0
      else 
        echo "EDGE CASE - DB EXISTS, BUT ENVAR IS NOT THERE OR FLAG FILE ALSO IS THERE WTF"
      fi
    fi 

    if [ "$?" != "0" ]; then
        echo ""
        echo "Seems like database failed to install."
        echo "Please check the error message above."
        echo ""
        return 1
    fi
```

So, now, if the API container nukes itself and it will just recreate the volumed file. But we probably don't even need the volume this way. Technically it will prevent the install script from running at all with the volumed file... But, as it is if the install script detects the DB is already installed and throws the special error code it can just move on instead of killing the process. 

The only caveat here is I had to modify a forked version the `api` to return an error code of `3` instead of `1` which appears to be the standard for all other error exit codes across that code base. It's a really simple change to enable the feature with a `docker-compose` ENVAR, but I am not sure how big of an issue the custom exit code is to the `api` team if you wanted to make it available as part of the suite.

Code from the `api` repo:
   ```
 private function handlePdoException(\PDOException $e)
    {
        $expected = "Duplicate entry 'directus_activity-id' for key 'idx_collection_field'";
        if ((int) $e->errorInfo[0] == 23000 && $e->errorInfo[2] == $expected) {
            echo "The Database for " . $this->directusPath . " had already been .\n";
            echo "You may use option -f to enforce the overwriting of existing data.";
            exit(3);
        } else {
            throw ($e);
        }
        exit(1);
    }
```

**Comment by zfalen** [comment]
Whoops. The only _other_ thing I did was modify the `Dockerfile` for the API base image to also install mysql tools for the DB ping command:
```
  apt-get update && \
  apt-get install default-mysql-client -y
```

**Comment by zfalen** [comment]
I can confirm that removing all references to the volumed `.dbInitialized` file works just fine.

**Comment by WoLfulus** [comment]
Hi @zfalen , I haven't been able to work these issues lately, but I just read everything again and we won't handle this kind of operational stuff since we need to make it free for the end user to use whenever works for them. Adding a solution like this will likely raise problems with people doing more complex stuff (like me in my projects), so I'll be agnostic here and let the users handle this themselves.

**Comment by zfalen** [comment]
Cool. At least this is here now so anyone else facing the same issue can see what we did to resolve it!
